### ISSUE TITLE
Audio "Fade over ... seconds" not possible to interrupt with Macro Stop

### ISSUE THREAD

**Issue by mutant-industries** [WarmUpTill/SceneSwitcher#417]
Macros have the option to fade in/out audio since [1.16.4](https://github.com/WarmUpTill/SceneSwitcher/releases/tag/1.16.4) and to stop running other macros since [1.17.2](https://github.com/WarmUpTill/SceneSwitcher/releases/tag/1.17.2).  The fade in/out duration can be configured (Fade over ... seconds). When e.g. 'Macro 1' triggers fade in/out of some audio source with 'Fade over ... seconds', then 'Macro 2' is triggered (during this time) with action Macro Stop 'Macro 1', then the fade in/out process is not stopped. Also when 'Macro 2' tries to fade in/out the same audio source as 'Macro 1', then it has no effect.

Checking / unchecking 'Run macro in parallel to other macros' on both macros has no effect here.

This can cause problems for example when doing cross-fade of audio levels using two macros. When the first one is triggered, the cross-fade in one direction starts. When the other one is triggered before the previous one finished, the cross-fade in oposite direction never starts.

My expectation is when stopping macro (with 'Macro Stop' action), any subsequent 'Audio Set source volume with Fade over ... seconds' actions of the stopped macro are also stopped and another 'Audio Set source volume with Fade over ... seconds' action on the same audio sources can be triggered immediately.

This happens with OBS 27.2.0-3-g2720e6d00 (64 bit) on fresh installation of Slackware 15.0 and latest Scene Switcher [1.17.2](https://github.com/WarmUpTill/SceneSwitcher/releases/tag/1.17.2). Nothing suspicious can be seen in log files.

**Comment by WarmUpTill** [comment]
> My expectation is when stopping macro (with 'Macro Stop' action), any subsequent 'Audio Set source volume with Fade over ... seconds' actions of the stopped macro are also stopped and another 'Audio Set source volume with Fade over ... seconds' action on the same audio sources can be triggered immediately.

Thanks for pointing this out!
I will have to rework the way the audio fade is handled to support this behaviour.

The reason for it behaving the way it currently does is that the audio fade predates the "run in background" option and "stop" action and at the time it was implemented I just moved the fading to a separate thread to allow this action to be performed during other actions.

**Comment by WarmUpTill** [comment]
A new build should be available here in a few minutes:
https://github.com/WarmUpTill/SceneSwitcher/actions/runs/1912123701
(You will have to be logged into GitHub to be able to download it)

This on should allow you to stop an ongoing fade using the Macro Stop action.

The new implementation now also allows to wait for a fade to complete:
![Fade](https://user-images.githubusercontent.com/19472752/156044978-a5323f0d-30ac-479a-b835-c4a828e6c83c.PNG)

Let me know if everything works as expected.

**Comment by mutant-industries** [comment]
The fix looks promising, however there is still one issue left. Let's have two audio sources and two macros, that look like this:
![ready_to_cam_macro](https://user-images.githubusercontent.com/9576228/156091207-cda896d9-cfd2-4701-9d59-168f17df9cc0.png)
and the other one here:
![cam_to_any_macro](https://user-images.githubusercontent.com/9576228/156091324-0ebfe96e-a183-4759-8467-578cf75ed28a.png)
As we can see the scene switch from ready scene to camera scene causes audio crossfade, where 'background audio' fades to 0% and 'OBS camera' fades to 100%, the crossfade takes 3.3 seconds. When the scene is switched back the crossfade in oposite direction is (or should be) started. Having situation like this several issues can be observed:
1. When first (or second) macro is triggered with 'Run macro' action before second (or first) macro is finished, the 'Macro Stop' action is executed properly, however nothing else happens. Only after the 'Run macro' is triggered once again the audio crossfade in oposite direction starts. The expected behavior is that on first invocation of 'Run macro' both actions are performed: the Stop of previous macro execution (the crossfade in one direction) and the crossfade in other direction. By adding Wait action for at least 0.15 seconds between macro Stop and the other audio fade trigger fixes the issue, but it is nothing but workaround. The situation then looks like this:
![cam_to_any_with_wait_macro](https://user-images.githubusercontent.com/9576228/156093401-13c13690-8f94-4117-ad27-ef202b94c119.png)
2. When triggering macros by actual Transition started event, the behavior is a bit different. When audio crossfade triggered by switch from ready scene to camera scene is triggered and then suddenly the other macro is triggered, then only the execution of the first macro is stopped and the crossfade in backwards direction is not started at all. The workaround again is using the Wait action.

**Comment by mutant-industries** [comment]
Regarding the crossfade I can think of one really nice to have feature. The only thing we can do now is fade from X% to Y% with duration of Z seconds. It would be nice to be able to set 'volume step [%] per second' instead of static duration of Z seconds. 

The use-case for this is again doing crossfade in one direction, then doing the interrupt at some random moment and starting the crossfade in the oposite direction. User might want to move the volume levels with static speed (e.g. 10% / second, so when moving from 0% to 50% the fade duration would be 5 seconds) instead of having every fade to take constant amount of time.

As a result let's say some macro with fade-in action from 0% to 60% with fade speed of 10% / second is running. This macro in interrupted by another macro (Stop action) after 3 seconds (so actual volume level is somewhere around 30%) and fade-out again with fade speed of 10% / second is triggered. Now the fade-out only takes 3 seconds.

**Comment by WarmUpTill** [comment]
Thanks for the feedback!
I reworked the macro stop functionality to block until the macro is actually stopped.
This in turn required a minor rework of the wait action.

A new build with these changes will be available here in a few minutes:
https://github.com/WarmUpTill/SceneSwitcher/actions/runs/1918148935
When using this build you should now no longer require the additional wait actions after stopping the macro.

Thanks for the suggestion of regarding the different approach of using the fade action.
I will look into it and report back if I have a first working version.
(Might get around to this on the weekend)

**Comment by mutant-industries** [comment]
With the latest build everything works as expected. Thank you, your plugin really makes lives a lot easier and thanks for exellent support!
